awx/main/tasks.py

```diff
@@ -520,18 +520,47 @@ def update_inventory_computed_fields(inventory_id, should_update_hosts=True):
     Signal handler and wrapper around inventory.update_computed_fields to
     prevent unnecessary recursive calls.
     '''
-    i = Inventory.objects.filter(id=inventory_id)
-    if not i.exists():
-        logger.error("Update Inventory Computed Fields failed due to missing inventory: " + str(inventory_id))
-        return
-    i = i[0]
-    try:
-        i.update_computed_fields(update_hosts=should_update_hosts)
-    except DatabaseError as e:
-        if 'did not affect any rows' in str(e):
-            logger.debug('Exiting duplicate update_inventory_computed_fields task.')
+    def _perform_update(inventory_id, should_update_hosts=True):
+        i = Inventory.objects.filter(id=inventory_id)
+        if not i.exists():
+            logger.error("Update Inventory Computed Fields failed due to missing inventory: " + str(inventory_id))
             return
-        raise
+        i = i[0]
+        try:
+            i.update_computed_fields(update_hosts=should_update_hosts)
+        except DatabaseError as e:
+            if 'did not affect any rows' in str(e):
+                logger.debug('Exiting duplicate update_inventory_computed_fields task.')
+                return
+            raise
+
+    compute_key = 'inventory_compute_{}'.format(inventory_id)
+    wait_key = 'inventory_wait_{}'.format(inventory_id)
+    with advisory_lock(compute_key, wait=False) as compute_lock:
+        if compute_lock:
+            _perform_update(inventory_id, should_update_hosts=True)
+        else:
+            with advisory_lock(wait_key, wait=False) as wait_lock:
+                if wait_lock:
+                    logger.debug('Obtained wait lock for inventory {}.'.format(inventory_id))
+                    with advisory_lock(compute_key, wait=True):
```

Comment:
This actually looks like a _third_ nested lock.  It's possible we've totally got the logic right such that it would never deadlock, but my alarm sirens are going off pretty loud on this PR.

Reply:
The first nested block, `with advisory_lock(compute_key, wait=False) as compute_lock:` I wrote, accepting the stated behavior of

https://github.com/Xof/django-pglocks

> wait (default True) -- If True (the default), the context manager will wait until the lock has been acquired before executing the content; in that case, it always returns True (unless a deadlock occurs, in which case an exception is thrown). If False, the context manager will return immediately even if it cannot take the lock, in which case it returns false. Note that the context body is always executed; the only way to tell in the wait=False case whether or not the lock was acquired is to check the returned value.

This line (the 3rd nested code) is the scenario where `False` was returned. I know that looks like it's nesting locks, but the lock is not held. Nested code here doesn't imply nested locks. That's probably the main thing I've failed to communicate here.